### PR TITLE
feat: add new on_failure parameter for healthchecks

### PR DESCRIPTION
**Pull request by noeleont**
as mentioned in #11 the on_failure parameter contains:

command:

Runs a command on the host with data about
the healthcheck and the failing container.

url:

Sends the failed check data to the specified url

### CODE REVIEW

**Comment by noeleont** [comment]
I'll move the changes to a `onFailure` function in `appjson/healthcheck.go` and reopen the PR.

**Comment by noeleont** [comment]
Since I'm not working with Go a lot can you tell me if this is a good approach or should I do it differently?

File: appjson/healthcheck.go [hunk 1]
```diff
@@ -129,6 +138,35 @@ func (h Healthcheck) Execute(container types.ContainerJSON, ctx HealthcheckConte
 	return h.executeUptimeCheck(container)
 }
 
+func (h Healthcheck) HandleFailure(errors []error) []error {
+	if h.OnFailure != nil {
```
Comment (josegonzalez) [inline appjson/healthcheck.go:142]: Return early if it's nil

File: commands/check.go [hunk 2]
```diff
@@ -247,6 +248,10 @@ func (c *CheckCommand) processHealthcheck(healthcheck appjson.Healthcheck, conta
 		if len(b) > 0 {
 			logger.Error(fmt.Sprintf("Error for healthcheck name='%s', output: %s", healthcheck.GetName(), strings.TrimSpace(string(b))))
 		}
+		err := healthcheck.HandleFailure(errs)
+		if err != nil {
```
Comment (josegonzalez) [inline commands/check.go:252]: This can be:

```go
if err := healthcheck.HandleFailures(errs); err != nil {
```

File: commands/check.go [hunk 3]
```diff
@@ -200,6 +200,7 @@ func (c *CheckCommand) Run(args []string) int {
 			hasErrors = true
 			err := resp.Errors[len(resp.Errors)-1]
 			logger.Error(fmt.Sprintf("Failure in name='%s': %s", resp.HealthcheckName, err.Error()))
+
```
Comment (josegonzalez) [inline commands/check.go:203]: We don't need the extra newline here.

File: appjson/healthcheck.go [hunk 4]
```diff
@@ -129,6 +138,35 @@ func (h Healthcheck) Execute(container types.ContainerJSON, ctx HealthcheckConte
...
+			if err != nil {
+				return []error{fmt.Errorf("failed to execute on failure command: %s", err)}
+			}
+		}
+		if len(h.OnFailure.Url) > 0 {
+			// NOTE: note sure if this is the correct way
+			err := errors[len(errors)-1]
+			data := map[string]interface{}{
+				"healthcheck_name": h.GetName(),
+				"errors":           err,
```
Comment (josegonzalez) [inline appjson/healthcheck.go:155]: We should probably send all errors.

File: appjson/healthcheck.go [hunk 5]
```diff
@@ -129,6 +138,35 @@ func (h Healthcheck) Execute(container types.ContainerJSON, ctx HealthcheckConte
...
+			}
+			json_data, err := json.Marshal(data)
+			if err != nil {
+				return []error{fmt.Errorf("failed to encode data as JSON: %s", err)}
+			}
+			_, err = http.Post(h.OnFailure.Url, "application/json", bytes.NewBuffer(json_data))
+			if err != nil {
```
Comment (josegonzalez) [inline appjson/healthcheck.go:162]:
```go
if err = http.Post(h.OnFailure.Url, "application/json", bytes.NewBuffer(json_data)); err != nil {
```

File: appjson/healthcheck.go [hunk 6]
```diff
@@ -129,6 +138,35 @@ func (h Healthcheck) Execute(container types.ContainerJSON, ctx HealthcheckConte
 	return h.executeUptimeCheck(container)
 }
 
+func (h Healthcheck) HandleFailure(errors []error) []error {
+	if h.OnFailure != nil {
+		if len(h.OnFailure.Command) > 0 {
+			cmd := exec.Command(h.OnFailure.Command[0], h.OnFailure.Command[1:]...)
+			err := cmd.Run()
+			if err != nil {
```
Comment (josegonzalez) [inline appjson/healthcheck.go:146]:
```go
if err := cmd.Run(); err != nil {
```

**Review by josegonzalez — CHANGES_REQUESTED** [review]

**Comment by josegonzalez** [comment]
Apologies for the late review, but this should be ready to try now :)

**Comment by josegonzalez** [comment]
Apologies for the late reply, but the code looks pretty good so far. Just needs a few fixes. Let me know if you'd like me to pick this up.

**Comment by noeleont** [comment]
No worries, I will pick it up in the next days, thanks for the review!

**Comment by noeleont** [comment]
@josegonzalez I implemented the changes :)

File: appjson/healthcheck.go [hunk 7]
```diff
@@ -129,6 +138,31 @@ func (h Healthcheck) Execute(container types.ContainerJSON, ctx HealthcheckConte
...
+	if len(h.OnFailure.Url) > 0 {
+		data := map[string]interface{}{
+			"healthcheck_name": h.GetName(),
+			"errors":           errors,
+		}
+		json_data, err := json.Marshal(data)
+		if err != nil {
+			return []error{fmt.Errorf("failed to encode data as JSON: %s", err)}
+		}
+		if _, err = http.Post(h.OnFailure.Url, "application/json", bytes.NewBuffer(json_data)); err != nil {
```
Comment (josegonzalez) [inline appjson/healthcheck.go:160]: Does `net/http.Post()` return an error on non-200?

File: appjson/healthcheck.go [hunk 8]
```diff
@@ -129,6 +138,31 @@ func (h Healthcheck) Execute(container types.ContainerJSON, ctx HealthcheckConte
 	return h.executeUptimeCheck(container)
 }
 
+func (h Healthcheck) HandleFailure(errors []error) []error {
```
Comment (josegonzalez) [inline appjson/healthcheck.go:141]: I think we should only need to return a single error ever from this func.

**Comment by josegonzalez** [comment]
Last two review comments, should be relatively easy to fix :)

File: appjson/healthcheck.go (hunk 7 above)
Comment (noeleont) [reply appjson/healthcheck.go]: I added the correct check

File: appjson/healthcheck.go (hunk 8 above)
Comment (noeleont) [reply appjson/healthcheck.go]: you're right, changed it

**Comment by noeleont** [comment]
I rebased to your branch and squashed the commits to streamline the history. Thanks for your help!

File: appjson/healthcheck.go [hunk 9]
```diff
@@ -25,16 +27,23 @@ type AppJSON struct {
...
+	Type         string     `json:"type,omitempty"`
+	Uptime       int        `json:"uptime,omitempty"`
+	Wait         int        `json:"wait,omitempty"`
+	OnFailure    *OnFailure `json:"onFailure,omitempty"`
+}
+
+type OnFailure struct {
+	Command       []string `json:"command,omitempty"`
+	Url           string   `json:"url,omitempty"`
+	OnNonzeroExit bool     `json:"onNonzeroExit,omitempty"`
```
Comment (josegonzalez) [inline appjson/healthcheck.go:46]: We can probably just remove this. Idk what I was thinking in the initial spec, but this is unused in your diff.

**Review by josegonzalez — CHANGES_REQUESTED** [review]
Minor nit and we'll be good to go

**Pull request merged**